scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/SortedBucketSource.java

```diff
@@ -667,14 +671,26 @@ long getOrSampleByteSize() {
           .sum();
     }
 
-    KeyGroupIterator<byte[], V> createIterator(int bucketId, int targetParallelism) {
+    KeyGroupIterator<byte[], V> createIterator(
+        int bucketId, int targetParallelism, PipelineOptions options) {
+      final SortedBucketOptions opts = options.as(SortedBucketOptions.class);
+      final int bufferSize = opts.getReadBufferSize();
+      final boolean backoff = opts.getReadBackoff();
       final List<Iterator<V>> iterators =
           mapBucketFiles(
               bucketId,
               targetParallelism,
-              file -> {
+              (partition, file) -> {
                 try {
-                  return fileOperations.iterator(file);
+                  if (backoff) {
+                    // we hit 10,000ms limit when partition = 23, 32,442ms cumulatively
+                    // 26.71min at partition = 180
+                    long millis = (long) Math.min(10000, Math.pow(1.5, partition));
+                    Thread.sleep(millis);
+                  }
+
+                  Iterator<V> iterator = fileOperations.iterator(file);
+                  return bufferSize > 0 ? new BufferedIterator<>(iterator, bufferSize) : iterator;
```

Comment:
hm since the default `bufferSize` is 10000, unless users explicitly set it to 0 when they upgrade to Scio 0.10.3 they'll get this change in behavior... did you have a chance to perf test it on some Dataflow jobs as well as DirectRunner?

the PR looks good otherwise! 👍